### ISSUE TITLE
Add unique descriptor to report requests to more easily distinguish them in testing and debugging.

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#7348]
## Feature Description

It would be useful to add some sort of human-readable descriptor to individual report requests that is sent as request param or header in order to be able to easily track the specific report down in devtools and so forth.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by zutigrm** [comment]
@benbowler How I see it, it would be helpful to have direct widget name included, the widget itself is usually telling enough details to identify it, for example KMW have distinctive names, as well as regular widgets, like `Dashboard Overall Page Metrics Widget`, perhaps we can automate it to include the widget title, but I don't think we need any extra information with it, since it is already provided in the request params. As to whether to add it automatically or manually, that would be more of an IB territory, what we can adjust here as AC part would be the end result of the label, I suggested the readable ones using their titles, but we can adjust that if needed.

**Comment by eugene-manuilov** [comment]
IB ✔

**Comment by techanvil** [comment]
Hey @zutigrm @benbowler, the AC was and the IB is still missing the crucial point that the report ID should be unique, as currently specced we'll end up with duplicate report IDs for widgets which request multiple reports, which is less useful for tracking.

It's still a good idea to formalise a prefix e.g. `module_widget-name_{widget|kmw}` as specced, but the ID should also include a unique component that has relevance to the individual report.

I've amended the AC, sending it back to you @zutigrm for an update to the IB.

**Comment by zutigrm** [comment]
Thanks @techanvil good shout. I updated the IB to include this

**Comment by techanvil** [comment]
Thanks @zutigrm! That's looking good, my one slight concern is that by making `ReportOptionName` optional, it's less future-proof in that, if a widget subsequently adds another report, we'll need to update the existing report ID to make it more distinct/descriptive.

It might be better to always include `ReportOptionName` to ensure report IDs can remain more static. WDYT?

**Comment by zutigrm** [comment]
@techanvil that's a good point. I updated it so it is not optional, and expanded on examples

**Comment by techanvil** [comment]
Thanks @zutigrm, that's looking better. However, the proposed approach still won't guarantee uniqueness - if we have multiple components in a widget that use the same variable name for the same report, their names will clash.

I'm not sure if there are any such cases in the codebase, but it would be good to take this into consideration and ensure we can always provide a unique identifier.

Actually, I've just realised there are some other cases we need to handle too - reports aren't always called directly from a widget, they can be called via a resolver which is used in multiple widgets, or possibly outside of the context of a widget - e.g. the gathering data reports:

https://github.com/google/site-kit-wp/blob/7c9f81c9da1c02b37f943ec22a986eb241beb02c/assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js#L164-L166

So, we need to think about how to accommodate these kind of cases too.

**Comment by zutigrm** [comment]
@techanvil good catch. Hm, that will throw us off the track when it comes to uniqueness, as the name construction so far was the closest one. 

I am seeing two remaining options to ensure uniqueness:
-  We assign each report a number ID, and track it in a new sheet, so we can never use same ID twice
   - Downside would be descriptiveness, as it would be lost 
- We write it in a explanatory sentence for each report, like `reportForRetrievingTitlesInTopPerformingPages`, `reportForCustomDimensionsGatheringDataWhichChecksForAvailability` etc

WDYT?

**Comment by techanvil** [comment]
Thanks, @zutigrm - however, I don't think we need to move that far away from the current proposed approach, just refine it. There is value to having common prefixes that we can use to filter groups of reports in the devtools.

I'd suggest broadening the scope for the prefix to allow for feature names (by feature name I mean something like `audience-segmentation`, `key-metrics`), as well as module names, and component, store & selector or other function names in addition to widgets.

So, it could have a structure of `{category}_{group}_{descriptor}`, where `category` is the module or feature, `group` is the widget/component/selector/function name, and `descriptor` is the report option name.

`{module|feature}_{widget|component|store:selector|other function}_{ReportOptionName}`

When providing a widget name, we could also have the option to include a subset of the group values to further qualify the report ID. Although, maybe this won't be needed initially. We could stick to the above to not overcomplicate things for now, but I'm thinking something like this if needs be:

`{module|feature}_{widget}_{component|store:selector|other function}_{ReportOptionName}`

I think this should be enough to provide good-enough uniqueness for our requirements, and we can always refine it further (for example we could specify that these values need to be provided as a string literal, and include a custom ESLint rule to enforce uniqueness).  What do you think?

**Comment by zutigrm** [comment]
Thanks @techanvil that could work. I tweaked the IB to follow those guidelines

**Comment by techanvil** [comment]
Thanks, @zutigrm, that LGTM!

IB ✅

**Comment by eugene-manuilov** [comment]
@benbowler, looks like there are merge conflicts. Could you please solve them?

**Comment by zutigrm** [comment]
This issues doesn't need to be QA:Eng, it can be tested regularly by a QA, @benbowler could you update the QAB with an example on how they can filter/locate the report ID? Afterwards we can remove the QA:Eng label

**Comment by benbowler** [comment]
Thanks @zutigrm. The report ID is included in the query string so they can be filtered in the Developer Tools network tab using the simple text search. eg. for `analytics-4_dashboard-all-traffic-widget-ga4_widget_pieArgs` on the dashboard:

<img width="1294" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMURSQ2c9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--74dc3c1bf2683b9b3b3833897723e46dea8e0770/Screenshot%202025-07-23%20at%2013.05.11.png" alt="Screenshot 2025-07-23 at 13.05.11.png" />

**Comment by kelvinballoo** [comment]
### QA Update ✅ 

Checked based on the QAB and used the string 'analytics-4_dashboard-all-traffic-widget-ga4_widget_pieArgs' and was able to filter that particular request.

<Details>

<img width="1265" height="282" alt="Image" src="https://github.com/user-attachments/assets/2b48a7e7-22c2-4720-b066-e32231d432df" />

</Details>

Checked also that reports have a reportID in the query params.

<Details>

<img width="1146" height="327" alt="Image" src="https://github.com/user-attachments/assets/92226409-9410-4622-9e33-7c92c7c972a4" />

</Details>
